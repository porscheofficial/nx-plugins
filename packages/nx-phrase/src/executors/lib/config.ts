import { existsSync, readFileSync } from "fs-extra"
import { resolve } from "path"

import { NonSensitiveArgs } from "./types"
import { ExecutorContext } from "@nrwl/devkit"
import { load } from "js-yaml"
import { PhraseClientConfig } from "./phrase"
import { BuildExecutorSchema } from "../build/schema"

interface ConfigFileFormat {
    phrase: {
        [appName: string]: {
            access_token: string
        }
    }
}

export interface InternalPhraseConfig extends NonSensitiveArgs {
    phraseClientConfig: PhraseClientConfig
}

function readConfig(context: ExecutorContext) {
    const projectName = context.projectName
    if (!projectName) {
        throw new Error("project name not set in context")
    }

    const configFileLocation = resolve(context.root, ".phrase.yml")

    if (!existsSync(configFileLocation)) {
        console.error(`Could not find configuration file at ${configFileLocation}`)
        process.exit(-1)
    }

    const configFileContent = load(readFileSync(configFileLocation, { encoding: "utf8" })) as ConfigFileFormat

    let error = false
    if (!configFileContent.phrase) {
        console.info(`Config file does not have 'phrase' root node.`)
        error = true
    }
    const configSection = configFileContent.phrase[projectName]
    if (!configSection) {
        console.info(`Config file has no configuration section for project '${projectName}'`)
        error = true
    }

    if (!configSection.access_token) {
        console.error(
            `'access_token' not set in .phrase.yml for translation target in project.json of project '${projectName}'.`
        )
        error = true
    }

    if (error) {
        process.exit(-1)
    }

    return configSection
}

function validateConfig(config: NonSensitiveArgs, projectName: string, requiredConfigProperties: string[]) {
    let error = false

    requiredConfigProperties
        .filter((requiredProperty) => !config[requiredProperty])
        .forEach((requiredProperty) => {
            console.info(
                `Configuration property '${requiredProperty}' not set for translation target in project.json of project '${projectName}'.`
            )
            error = true
        })

    return !error
}

export function getConfig(
    options: BuildExecutorSchema,
    context: ExecutorContext,
    requiredConfigurationProperties: string[]
): InternalPhraseConfig {
    if (!context.projectName) {
        throw new Error("project name not set in context")
    }

    const {
        projectName,
        workspace: { projects },
    } = context
    const { sourceRoot, root } = projects[projectName]

    const rawConfig = readConfig(context)
    // get output first, so it does not get overwritten
    const output = resolve(
        root,
        // use from project.json -> config file -> fallback
        options.output ?? "./translations"
    )
    const config = {
        output,
        branch: options.branch,
        fileFormat: options.fileFormat ?? "react_simple_json",
        ignoreGlob: options.ignoreGlob ?? "**/*.d.ts",
        projectId: options.projectId,
        sourceRoot: options.sourceRoot ?? sourceRoot,
        sourceGlob: options.sourceGlob ?? "**/*.{ts,tsx}",
        uploadLanguageId: options.uploadLanguageId || "default",
        sourceKeyTransformer: options.sourceKeyTransformer,
        phraseKeyTransformer: options.phraseKeyTransformer,
        sourceKeyFilter: options.sourceKeyFilter,
        phraseKeyFilter: options.phraseKeyFilter,
        useFallbackLocale: options.useFallbackLocale,
        useSourceLocaleAsFallback: options.useSourceLocaleAsFallback,
    } as NonSensitiveArgs

    if (!validateConfig(config, projectName, requiredConfigurationProperties)) {
        throw new Error("Invalid configuration")
    }

    const phraseClientConfig: PhraseClientConfig = { token: rawConfig.access_token }

    return {
        phraseClientConfig,
        ...config,
    }
}
