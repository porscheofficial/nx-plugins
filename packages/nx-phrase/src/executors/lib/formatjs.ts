import { spawn } from "child_process"

export async function extract({
    sourceGlob,
    ignoreGlob,
    sourceRoot,
    outputFile,
}: {
    sourceGlob: string
    ignoreGlob?: string
    sourceRoot: string
    outputFile: string
}) {
    return new Promise<void>((resolve, reject) => {
        const extract = spawn(
            "./node_modules/.bin/formatjs",
            [
                "extract",
                `${sourceRoot}/${sourceGlob}`, // `./${projectRoot}/src/**/*.{ts,tsx}`,
                ...(ignoreGlob
                    ? [
                          "--ignore",
                          `${sourceRoot}/${ignoreGlob}`, //`./${projectRoot}/src/**/*.d.ts`,
                      ]
                    : []),
                "--out-file",
                outputFile,
                "--extract-source-location",
            ],
            { stdio: "inherit" }
        )

        extract.on("exit", (code) => {
            if (code === 0) {
                resolve()
            } else {
                reject()
            }
        })
    })
}

export async function compile({ inputFile, outputFile }: { inputFile: string; outputFile: string }) {
    return new Promise<void>((resolve, reject) => {
        const compile = spawn("./node_modules/.bin/formatjs", ["compile", inputFile, "--out-file", outputFile], {
            stdio: "inherit",
        })

        compile.on("exit", (code) => {
            if (code === 0) {
                resolve()
            } else {
                reject()
            }
        })
    })
}
