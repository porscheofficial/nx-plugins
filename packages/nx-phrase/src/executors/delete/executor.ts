import { ExecutorContext } from "@nrwl/devkit"
import { existsSync } from "fs"

import { getConfig, InternalPhraseConfig } from "../lib/config"
import { deleteKeys } from "../lib/delete"
import { NonSensitiveArgs } from "../lib/types"

const requiredConfigProperties = ["projectId", "inputFile"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, requiredConfigProperties)
    const { projectName } = context

    // get path from config / cli param?
    if (!existsSync(options.inputFile)) {
        throw new Error(`input file "${options.inputFile}" does not exist.`)
    }

    // load file
    await deleteKeys(options.inputFile, config)

    return { success: true }
}
