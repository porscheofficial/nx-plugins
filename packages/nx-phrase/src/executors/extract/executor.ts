import { ExecutorContext } from "@nx/devkit"

import { getConfig, InternalPhraseConfig } from "../../lib/config"
import { NonSensitiveArgs } from "../../lib/types"
import { extract } from "../../lib/extract"

export const pullRequiredConfigs = ["output"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, pullRequiredConfigs)

    await extract(config, context)

    console.log("Done.")

    return { success: true }
}
