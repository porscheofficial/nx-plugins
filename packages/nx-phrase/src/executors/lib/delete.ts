import { readFileSync } from "fs"
import { InternalPhraseConfig } from "./config"
import { PhraseClient } from "./phrase"

export async function deleteKeys(file: string, config: InternalPhraseConfig): Promise<void> {
    const keysToDelete = JSON.parse(readFileSync(file).toString())

    // ask for confirmation?
    const phrase = new PhraseClient(config.phraseClientConfig)
    for (const key of keysToDelete) {
        // resolve keyIds for key name
        const keyIds = await phrase.resolveKeyIdByName({
            projectId: config.projectId,
            keyName: key,
            branch: config.branch,
        })
        // delete keyIds
        for (const keyId of keyIds) {
            try {
                await phrase.deleteKey({ projectId: config.projectId, keyId, branch: config.branch })
            } catch (e) {
                console.log(`Could not delete key with id '${keyId}' for name '${key}', ${e}`)
            }
        }
    }
}
