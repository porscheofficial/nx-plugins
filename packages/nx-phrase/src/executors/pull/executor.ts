import { ExecutorContext } from "@nrwl/devkit"

import { getConfig, InternalPhraseConfig } from "../lib/config"
import { pull } from "../lib/pull"
import { NonSensitiveArgs } from "../lib/types"

export const pullRequiredConfigs = ["projectId", "output"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, pullRequiredConfigs)

    await pull(config)

    console.log("Done.")

    return { success: true }
}
