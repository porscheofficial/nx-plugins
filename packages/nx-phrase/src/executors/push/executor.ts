import { ExecutorContext } from "@nx/devkit"

import { getConfig, InternalPhraseConfig } from "../../lib/config"
import { push } from "../../lib/push"
import { NonSensitiveArgs } from "../../lib/types"

export const pushRequiredConfigs = ["projectId", "uploadLanguageId"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, pushRequiredConfigs)

    await push(config, context)

    console.log("Done.")

    return { success: true }
}
