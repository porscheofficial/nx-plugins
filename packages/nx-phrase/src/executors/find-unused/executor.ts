import { ExecutorContext } from "@nrwl/devkit"
import { existsSync, writeFileSync } from "fs"
import { resolve } from "path"
import { prepareOutput } from "../../utils"

import { getConfig, InternalPhraseConfig } from "../lib/config"
import { PullHelper } from "../lib/pull"
import { extractTranslations } from "../lib/push"
import { NonSensitiveArgs } from "../lib/types"

const requiredConfigProperties = ["projectId", "output"]

function defaultTransformKeyFn(translationKey: string) {
    return translationKey
}
type transformTranslationKeyFn = typeof defaultTransformKeyFn

function defaultFilterKeyFn(translationKey: string) {
    return !!translationKey
}
type filterKeyFn = typeof defaultFilterKeyFn

async function loadTransformer<T = transformTranslationKeyFn>(path: string) {
    if (path) {
        if (existsSync(path)) {
            const filterFunction = (await import(path)).default as T
            if (typeof filterFunction === "function") {
                return filterFunction
            } else {
                throw new Error(`${path} does not export a function as default`)
            }
        } else {
            throw new Error(`${path} does not exist`)
        }
    }
}

async function getKeysFromSource(config: InternalPhraseConfig, outputFilePath: string) {
    // extracts current translations
    const compiledTranslationsFilePath = await extractTranslations(config, outputFilePath)
    const compiledTranslationKeys = Object.keys((await import(compiledTranslationsFilePath)).default)

    return compiledTranslationKeys
}

async function getKeysFromPhrase(config: InternalPhraseConfig) {
    const pull = new PullHelper({ ...config, fileFormat: "react_simple_json" })

    // downloads available translations
    const locales = await pull.listLocales()
    const localeToRawTranslations = await pull.downloadTranslations(locales)
    const keysInPhrase = new Set<string>()
    Object.keys(localeToRawTranslations).forEach((locale) =>
        Object.keys(JSON.parse(localeToRawTranslations[locale])).forEach((key) => keysInPhrase.add(key))
    )

    // use set for deduplication
    const keysInPhraseList = [...keysInPhrase]

    return keysInPhraseList
}

function applyTransform(input: string[], transformer: transformTranslationKeyFn) {
    // apply optional filter (use set for deduplication)
    const set = new Set<string>()
    input.map(transformer).forEach((key) => set.add(key))
    return [...set]
}

function filterList(input: string[], filter: filterKeyFn) {
    return input.filter(filter)
}

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, requiredConfigProperties)
    const { projectName } = context

    // extract and prepare keys from source code
    const outputPath = await prepareOutput({
        projectRoot: context.workspace.projects[context?.projectName]?.root ?? context.root,
        subfolder: "unused",
        workingDirectory: options.workingDirectory,
    })
    let sourceTranslationKeys
    {
        const sourceKeyFilter = options.sourceKeyFilter
            ? await loadTransformer<filterKeyFn>(resolve(context.root, options.sourceKeyFilter))
            : defaultFilterKeyFn

        const sourceKeyTransformer = options.sourceKeyTransformer
            ? await loadTransformer(resolve(context.root, options.sourceKeyTransformer))
            : defaultTransformKeyFn

        const allSourceTranslationKeys = await getKeysFromSource(config, outputPath)
        sourceTranslationKeys = filterList(allSourceTranslationKeys, sourceKeyFilter)

        if (allSourceTranslationKeys.length !== sourceTranslationKeys.length) {
            const bogusKeys = allSourceTranslationKeys.filter((key) => !sourceTranslationKeys.includes(key))
            const sourceBugKeysFilename = `${outputPath}/${projectName}.filtered-source-keys.json`
            writeFileSync(sourceBugKeysFilename, JSON.stringify(bogusKeys, null, 2), {
                flag: "w",
            })
            console.log(`Keys filtered from source written to: ${sourceBugKeysFilename}`)
        }

        sourceTranslationKeys = applyTransform(sourceTranslationKeys, sourceKeyTransformer)
    }

    // extract and prepare keys from phrase
    let phraseTranslationKeys
    {
        const phraseKeyFilter = options.phraseKeyFilter
            ? await loadTransformer<filterKeyFn>(resolve(context.root, options.phraseKeyFilter))
            : defaultFilterKeyFn

        const phraseKeyTransformer = options.phraseKeyTransformer
            ? await loadTransformer(resolve(context.root, options.phraseKeyTransformer))
            : defaultTransformKeyFn

        const allPhraseTranslationKeys = await getKeysFromPhrase(config)
        phraseTranslationKeys = filterList(allPhraseTranslationKeys, phraseKeyFilter)

        if (allPhraseTranslationKeys.length !== phraseTranslationKeys.length) {
            const bogusKeys = allPhraseTranslationKeys.filter((key) => !phraseTranslationKeys.includes(key))
            const phraseBugKeysFilename = `${outputPath}/${projectName}.filtered-phrase-keys.json`
            writeFileSync(phraseBugKeysFilename, JSON.stringify(bogusKeys, null, 2), {
                flag: "w",
            })
            console.log(`Keys filtered from phrase written to: ${phraseBugKeysFilename}`)
        }

        phraseTranslationKeys = applyTransform(phraseTranslationKeys, phraseKeyTransformer)
    }

    // compare available and current translations
    const notUploadedYet = sourceTranslationKeys.filter((key) => !phraseTranslationKeys.includes(key))
    const unusedKeys = phraseTranslationKeys.filter((key) => !sourceTranslationKeys.includes(key))

    // tell difference
    const unusedKeysFilename = `${outputPath}/${projectName}.unused-keys.json`
    writeFileSync(unusedKeysFilename, JSON.stringify(unusedKeys, null, 2), {
        flag: "w",
    })
    console.log(`Unused keys written to: ${unusedKeysFilename}`)

    const pendingUploadKeysFilename = `${outputPath}/${projectName}.pending-upload-keys.json`
    writeFileSync(pendingUploadKeysFilename, JSON.stringify(notUploadedYet, null, 2), {
        flag: "w",
    })
    console.log(`Keys pending upload written to: ${pendingUploadKeysFilename}`)

    // console.log(`${context.targetName} ${context.configurationName}`)

    console.log("Done.")

    return { success: true }
}
