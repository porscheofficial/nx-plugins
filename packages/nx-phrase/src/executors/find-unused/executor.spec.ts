import { resolve } from "path"

import { ExecutorContext } from "@nrwl/devkit"
import nock from "nock"

import executor from "./executor"
import { NonSensitiveArgs } from "../lib/types"

const options: Partial<NonSensitiveArgs> = {
    phraseKeyTransformer: "./filter.js",
    sourceKeyTransformer: "./filter.js",
}

const TEST_ASSETS_DIR = resolve(__dirname, "../../../test")

function nockForProject(projectId = "project_id") {
    nock("https://api.phrase.com/v2")
        .get(`/projects/${projectId}/locales`)
        .matchHeader("Authorization", /token .*/)
        .query({ per_page: 20 })
        .replyWithFile(200, `${TEST_ASSETS_DIR}/localesListResponse.json`)
        .get(/\/projects\/[^/]+\/locales\/[^/]+\/download/)
        .matchHeader("Authorization", /token .*/)
        .query({ file_format: "react_simple_json" })
        .twice()
        .replyWithFile(200, `${TEST_ASSETS_DIR}/localeDownloadResponse.json`)
}

describe("find unused", () => {
    const context = {
        root: TEST_ASSETS_DIR,
        projectName: "test_app",
        workspace: { projects: { test_app: { root: TEST_ASSETS_DIR, sourceRoot: TEST_ASSETS_DIR } } } as unknown,
    } as ExecutorContext

    afterEach(() => {
        expect(nock.isDone()).toBe(true)
    })

    it("finds unused translations", async () => {
        nockForProject(options.projectId)

        const output = await executor(options, context)
        expect(output.success).toBe(true)
    })

    it("can override configuration options via project.json", async () => {
        nockForProject("other_project")
        const output = await executor({ ...options, projectId: "other_project" }, context)
        expect(output.success).toBe(true)
    })
})
