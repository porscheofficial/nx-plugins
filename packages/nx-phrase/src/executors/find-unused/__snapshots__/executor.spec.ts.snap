// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find unused filters before applying transformers: keys downloaded from phrase and FILTERED that are NOT used in the source (anymore) 1`] = `
Array [
  "a.c.d.e",
  "a.d.e.c",
  "o.l.e.o",
  "K.u.i.l",
]
`;

exports[`find unused filters before applying transformers: keys from the source and FILTERED that are NOT uploaded in phrase yet 1`] = `
Array [
  "sampleId",
]
`;

exports[`find unused filters keys: keys from phrase that are NOT valid according to the filter function 1`] = `
Array [
  "K.u.i.l[",
  "K.u.i.l]",
  "K.u.i.l[]",
  "K.u.i.l[,a]",
  "K.u.i.l[a,]",
  "K.u.i.l[ ]",
  "K.u.i.l[ , a ]",
  "K.u.i.l[ a , ]",
]
`;

exports[`find unused filters keys: keys from source that are NOT valid according to the filter function 1`] = `
Array [
  "a.b.c.d{",
]
`;

exports[`find unused filters keys: keys that exist in phrase more than once 1`] = `
Array [
  "a.b.c.d",
]
`;

exports[`find unused transforms keys: keys downloaded from phrase that are NOT used in the source (anymore) 1`] = `
Array [
  "a.c.d.e",
  "a.d.e.c",
  "o.l.e.o",
  "K.u.i.l",
  "K.u.i.l[",
  "K.u.i.l]",
]
`;

exports[`find unused transforms keys: keys from the source that are NOT uploaded in phrase yet 1`] = `
Array [
  "a.b.c.d{",
  "sampleId",
]
`;
